xonsh/history/json.py

```diff
@@ -15,53 +15,70 @@
 import xonsh.xoreutils.uptime as uptime
 
 
-def _xhj_gc_commands_to_rmfiles(hsize, files):
-    """Return the history files to remove to get under the command limit."""
-    rmfiles = []
+def _xhj_gc_commands_to_rmfiles(hsize, files) -> ([], int):
+    """Return number of units and list of history files to remove to get under the limit,
+
+    Parameters:
+    -----------
+    hsize (int):  units of history, # of commands in this case.
+    files ((mod_ts, num_commands, path)[], fsize): history files, sorted oldest first.
+
+    Returns:
+    --------
+    hsize_removed (int):  units of history to be removed
+    rm_files ((mod_ts, num_commands, path, fsize)[]): list of files to remove.
+    """
     n = 0
     ncmds = 0
-    for ts, fcmds, f in files[::-1]:
-        if fcmds == 0:
-            # we need to make sure that 'empty' history files don't hang around
-            rmfiles.append((ts, fcmds, f))
+    for _, fcmds, f, _ in files[::-1]:
+        # `files` comes in with empty files included (now), don't need special handling to gc them here.
+
         if ncmds + fcmds > hsize:
             break
         ncmds += fcmds
         n += 1
-    rmfiles += files[:-n]
-    return rmfiles
+
+    cmds_removed = 0
+    for _, fcmds, f, _ in files[:-n]:
```

Comment:
maybe assign `files[:-n]` to a variable rather than creating the list twice, like `rm_files = files[:-n]`